File: leetcode/py/searchMatrix.py
class Solution:
    def helper(self, matrix, x1, y1, x2, y2, target):
        if x1 > x2 or y1 > y2:
            return False
        xmid = int((x1 + x2)/2)
        ymid = int((y1 + y2)/2) 
        if target == matrix[xmid][ymid]:
            return True
        elif target < matrix[xmid][ymid] :
            xx2 = xmid - 1
            yy2 = ymid - 1
            return (self.helper(matrix, x1, y1, xx2, y2, target) or
                    self.helper(matrix, x1, y1, x2, yy2, target) )
        else:
            xx1 = xmid + 1
            yy1 = ymid + 1   
            return (self.helper(matrix, xx1, y1, x2, y2, target) or 
                    self.helper(matrix, x1, yy1, x2, y2, target) )
        
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        hlen = len(matrix)
        if hlen == 0:
            return False
        wlen = len(matrix[0])
        if wlen == 0:
            return False
        return self.helper(matrix, 0, 0, hlen-1, wlen-1, target)
        

File: leetcode/cpp/searchMatrix.cc
class Solution {
public:
    bool searchHelper(vector<vector<int>> & matrix, int xs, 
                      int ys, int xe, int ye, int target){
        if( xs > xe || ys > ye) return false;
        
        int xmid = (xs+xe)/2;
        int ymid = (ys+ye)/2;
        
        if(target == matrix[xmid][ymid] ) return true;
        else{
            if(target < matrix[xmid][ymid]){
                return (searchHelper(matrix, xs, ys, xmid-1 , ye, target) 
                        || searchHelper(matrix, xs, ys, xe, ymid-1, target));
            }
            else{
                return (searchHelper(matrix, xmid+1, ys, xe , ye, target) 
                        || searchHelper(matrix, xs, ymid+1, xe, ye, target));
            }
        }
    }
    
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int hlen = matrix.size();
        if( hlen == 0 ) return false;
        int wlen = matrix[0].size();
        if( wlen == 0 ) return false;
        return searchHelper(matrix, 0, 0, hlen-1, wlen-1, target);
    }
};
